plugin/trino-hive/src/main/java/io/trino/plugin/hive/aws/athena/projection/DateProjectionFactory.java

```diff
@@ -0,0 +1,202 @@
+        if (!leftBound.get().isBefore(rangeBound.get())) {
+            throw invalidRangeProperty(columnName, dateFormatPattern);
+        }
+
+        int interval = getProjectionPropertyValue(columnProperties, COLUMN_PROJECTION_INTERVAL, Integer.class::cast).orElse(1);
+        ChronoUnit intervalUnit = getProjectionPropertyValue(columnProperties, COLUMN_PROJECTION_INTERVAL_UNIT, ChronoUnit.class::cast)
+                .orElseGet(() -> resolveDefaultChronoUnit(columnName, dateFormatPattern));
+
+        if (!DATE_PROJECTION_INTERVAL_UNITS.contains(intervalUnit)) {
+            throw invalidProjectionException(
+                    columnName,
+                    format(
+                            "Property: '%s' value '%s' is invalid. Available options: %s",
+                            COLUMN_PROJECTION_INTERVAL_UNIT,
+                            intervalUnit,
+                            DATE_PROJECTION_INTERVAL_UNITS));
+        }
+
+        return new DateProjection(columnName, dateFormat, leftBound, rangeBound, interval, intervalUnit);
+    }
+
+    private ChronoUnit resolveDefaultChronoUnit(String columnName, String dateFormatPattern)
+    {
+        if (dateFormatPattern.contains("S") || dateFormatPattern.contains("s")
+                || dateFormatPattern.contains("m") || dateFormatPattern.contains("H")) {
+//            When the provided dates are at single-day or single-month precision.
+            throw invalidProjectionException(
+                    columnName,
+                    format(
+                            "Property: '%s' needs to be set when provided '%s' is less that single-day precision. Interval defaults to 1 day or 1 month, respectively. Otherwise, interval is required",
+                            COLUMN_PROJECTION_INTERVAL_UNIT,
+                            COLUMN_PROJECTION_FORMAT));
+        }
+        if (dateFormatPattern.contains("d")) {
+            return DAYS;
+        }
+        return MONTHS;
+    }
+
+    private Supplier<Instant> parseDateRangerBound(String columnName, String value, SimpleDateFormat dateFormat)
+    {
+        if (value.contains("NOW")) {
+            Matcher matcher = DATE_RANGE_BOUND_EXPRESSION_PATTERN.matcher(value);
+            if (matcher.matches()) {
+                String operator = matcher.group(2);
+                String multiplierString = matcher.group(3);
+                String unitString = matcher.group(4);
+                if (nonNull(operator) && nonNull(multiplierString) && nonNull(unitString)) {
+                    unitString = unitString.toUpperCase(Locale.ENGLISH);
+                    return new DateExpressionBound(
+                            Integer.parseInt(multiplierString),
+                            ChronoUnit.valueOf(unitString + "S"),
+                            operator.charAt(0) == '+');
+                }
+                if (value.trim().equals("NOW")) {
+                    return Instant::now;
```

Comment:
generally using `Instant.now()` feels wrong. It should be a fixed value through query execution (if we end up calling this twice while executing the query we should get the same values). We should use instant of when query has been started. It is available in session - but I think it may be hard to get it here. Plese see what we can do.

Reply:
Yes we don't have access to session. We would need to implement some kind of QuerySessionService which would allow to access session by queryId and it will be added there during query start and then may be injected to any location, and that way anyone anywhere will be able to access session by queryId.